cumulusci/tasks/salesforce/sourcetracking.py

```diff
@@ -0,0 +1,234 @@
+        for namespace in self.options["exclude_namespace"]:
+            self._exclude.append("CustomField: .*\.{}__.*__c".format(namespace))
+            self._exclude.append("CompactLayout: .*\.{}__.*__c".format(namespace))
+        if self.project_config.project__source__ignore:
+            self._exclude.extend(self.project_config.project__source__ignore)
+
+        self._load_snapshot()
+
+    @property
+    def _snapshot_path(self):
+        parent_dir = os.path.join(".cci", "snapshot")
+        if not os.path.isdir(parent_dir):
+            os.makedirs(parent_dir)
+        return os.path.join(parent_dir, "{}.json".format(self.org_config.name))
+
+    def _load_snapshot(self):
+        self._snapshot = {}
+        if os.path.isfile(self._snapshot_path):
+            with open(self._snapshot_path, "r") as f:
+                self._snapshot = json.load(f)
+
+    def _get_changes(self):
+        changes = self.tooling.query(
+            "SELECT MemberName, MemberType, RevisionNum FROM SourceMember WHERE IsNameObsolete=false"
+        )
+        return changes
+
+    def _store_snapshot(self):
+        with open(self._snapshot_path, "w") as f:
+            f.write(json.dumps(self._snapshot))
+
+    def _run_task(self):
+        changes = self._get_changes()
+        if changes["totalSize"]:
+            self.logger.info(
+                "Found {} changed components in the scratch org:".format(
+                    changes["totalSize"]
+                )
+            )
+
+        filtered = self._filter_changes(changes)
+        for change in filtered:
+            self.logger.info("{MemberType}: {MemberName}".format(**change))
+        if not filtered:
+            self.logger.info("Found no changes.")
+
+        ignored = len(changes["records"]) - len(filtered)
+        if ignored:
+            self.logger.info(
+                "Ignored {} changed components in the scratch org:".format(ignored)
+            )
+
+        if self.options["snapshot"]:
+            self.logger.info("Storing snapshot of changes")
+            self._store_snapshot()
+
+    def _filter_changes(self, changes):
+        filtered = []
+        for change in changes["records"]:
+            mdtype = change["MemberType"]
+            name = change["MemberName"]
+            full_name = "{}: {}".format(mdtype, name)
+            if self.options["include"] and not any(
+                re.search(s, full_name) for s in self.options["include"]
+            ):
+                continue
+            if any(re.search(s, full_name) for s in self._exclude):
+                continue
+            revnum = self._snapshot.get(mdtype, {}).get(name)
+            if revnum and revnum == change["RevisionNum"]:
+                continue
+            filtered.append(change)
+
+            self._snapshot.setdefault(mdtype, {})[name] = change["RevisionNum"]
+
+        return filtered
+
+
+retrieve_changes_task_options = ListChanges.task_options.copy()
```

Comment:
I'm open to that.  There are a number of places we use this pattern though.  It might be best to do a dedicate branch on changing that style across all the tasks that use it.

Reply:
This is actually a case where I think repetition is better. There's no reason we need the same objects (in terms of object identity) within the task_options dict, and doing it this way makes it harder to see at a glance what they are when reading the code.